Addition of pixelClusterTagInfos to AOD and miniAOD

zucchett opened the pull request:
#### PR description:

This pull request defines a new TagInfo object (PixelClusterTagInfo), produced from the RECO format starting from the siPixelCluster collection, propagated to AOD, and embedded in miniAOD. The TagInfo contains a very compressed information of the pixel clusters geometrically matched to an AK4 jet, consisting of a char for each 4 pixel layers times 6 jet-cluster association cones, plus four unsigned integers for the collected charge. This information is added only to jets with pT > 100 GeV, and is critical for novel studies of b-tagging at high pT without relying on tracks, whose reconstruction efficiency decreases in very energetic jets. The pixel cluster information can be provided to a standalone tagger or (in the future) integrated in combined b-tagging algorithms. Preliminary studies performed on very limited samples estimate that a pixel-based tagger provides a better performance than the DeepCSV for jet pT > 1.2 TeV.

Latest presentations:
https://indico.cern.ch/event/828712/contributions/3468910/attachments/1869375/3075401/Sommerhalder_Zucchetta.pdf
https://indico.cern.ch/event/830302/contributions/3483708/attachments/1871179/3079451/190628_Zucchetta.pdf
https://indico.cern.ch/event/833434/contributions/3495708/attachments/1877898/3093199/190710_Zucchetta.pdf

#### PR validation:

The basic checks have been successfully run (code-checks, runTheMatrix). The only exception is represented by two jobs in runTheMatrix that run on already-produced AOD (136.7611* and 136.8311*) that fail because the needed collections are not found. This is understandable because neither the basic siPixelCluster nor the pixelClusterTagInfo collections are present in AODs by default; the latter collection is added to AODs with this PR.
EDIT: Era modifiers have been introduced to exclude pixelClusterTagInfos from 80X and 94X miniAOD productions, avoiding the crash.


@msommerh as developer, and BTV conveners @mverzett , @kskovpen, @XavierAtCERN, @smoortga are also involved and interested in these developments.

zucchett committed:
First commit of the PixelClusterTagInfo class

zucchett committed:
First commit of the PixelClusterTagInfo class

zucchett committed:
Added PixelClusterTagInfo and Collection class info

zucchett committed:
Added PixelClusterTagInfo and Collection class info

zucchett committed:
First commit of the new PixelCluster subpackage

zucchett committed:
First commit of the new PixelCluster subpackage

zucchett committed:
Added default cfg

zucchett committed:
Added default cfg

zucchett committed:
Added readout of pixel cluster info

zucchett committed:
Added readout of pixel cluster info

zucchett committed:
Updated and added parameters as number of layers

zucchett committed:
Updated and added parameters as number of layers

zucchett committed:
Added PixelClusterProperties struct definition

zucchett committed:
Added PixelClusterProperties struct definition

zucchett committed:
Added new int to Properties and change type to some Data

zucchett committed:
Added new int to Properties and change type to some Data

zucchett committed:
Added cluster matching; first fully working code

zucchett committed:
Added cluster matching; first fully working code

zucchett committed:
updated inputs and moved to dynamic allocation

zucchett committed:
updated inputs and moved to dynamic allocation

zucchett committed:
Set static allocation and initialization for data struct

zucchett committed:
Set static allocation and initialization for data struct

zucchett committed:
Switched to one single jet collection

zucchett committed:
Switched to one single jet collection

zucchett committed:
Added pixelCluster_cff to run on three different jet collections

zucchett committed:
Added pixelCluster_cff to run on three different jet collections

zucchett committed:
Re-added include after rebase

zucchett committed:
Re-added include after rebase

zucchett committed:
Initialize all values to zero in struct

zucchett committed:
Initialize all values to zero in struct

zucchett committed:
Removed processes for different jet collections

zucchett committed:
Removed processes for different jet collections

zucchett committed:
Added pixel cluster tag infos to jetTools

zucchett committed:
Added pixel cluster tag infos to jetTools

zucchett committed:
Added pixel cluster tag infos to bTagging_cff and added fake discriminator

zucchett committed:
Added pixel cluster tag infos to bTagging_cff and added fake discriminator

zucchett committed:
Added pixel cluster ro RecoBTag sequences

zucchett committed:
Added pixel cluster ro RecoBTag sequences

zucchett committed:
Added pixel cluster EventSetup_cff

zucchett committed:
Added pixel cluster EventSetup_cff

zucchett committed:
Split PixelClusterTagInfoProducer in .cc and .h files

zucchett committed:
Split PixelClusterTagInfoProducer in .cc and .h files

zucchett committed:
Forced embedding of pixelClusterTagInfos in slimmedJets

zucchett committed:
Forced embedding of pixelClusterTagInfos in slimmedJets

zucchett committed:
Removed pixelCluster from jetTools to prevent crashes

zucchett committed:
Removed pixelCluster from jetTools to prevent crashes

zucchett committed:
Embed pixelClusterTagInfos to patJets and keep them up to slimmedJets in miniAOD production

zucchett committed:
Embed pixelClusterTagInfos to patJets and keep them up to slimmedJets in miniAOD production

zucchett committed:
Cleanup code of pixelClusterTagInfoProducer class; added back fillDescriptions

zucchett committed:
Cleanup code of pixelClusterTagInfoProducer class; added back fillDescriptions

zucchett committed:
Removed unused dependencies and CMSSW version

zucchett committed:
Removed unused dependencies and CMSSW version

zucchett committed:
Replaced inputTag full lists with append in miniAOD cfg

zucchett committed:
Replaced inputTag full lists with append in miniAOD cfg

zucchett committed:
Added pixelClusterTagInfos to miniAOD content

zucchett committed:
Added pixelClusterTagInfos to miniAOD content

zucchett committed:
PixelClusterTagInfoProducer updated with Code checks modifications

zucchett committed:
PixelClusterTagInfoProducer updated with Code checks modifications

zucchett committed:
Merge branch 'PixelClusterTagInfo' of https://github.com/zucchett/cmssw into PixelClusterTagInfo

zucchett committed:
Remove the embedding of all other tagInfos, as the corresponding collections are not included and crash the nanoAOD sequence

zucchett committed:
Added sequence modifier for FastSim to avoid siPixelCluster product not found

zucchett committed:
Prevent running the pixelCluster sequence when creating miniAODs from AOD produced with FastSim

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11371

- This PR adds an extra 144KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11371/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11371/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

zucchett committed:
Modified indentation according to CMS code-format rules

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11373

- This PR adds an extra 152KB to repository

cmsbuild commented:
A new Pull Request was created by  @zucchett (Alberto Zucchetta) for master.

It involves the following packages:

DataFormats/BTauReco
PhysicsTools/PatAlgos
RecoBTag/Configuration
RecoBTag/PixelCluster


The following packages do not have a category, yet:

RecoBTag/PixelCluster
Please create a PR for https://github.com/cms-sw/cms-bot/blob/master/categories_map.py to assign category

@perrotta, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@emilbols, @gouskos, @rappoccio, @HeinerTholen, @peruzzim, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @acaudron, @jdolen, @ferencek, @rovere, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @mariadalfonso, @clelange, @JyothsnaKomaragiri, @mverzett, @gpetruc, @andrzejnovak, @pvmulder this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
> The only exception is represented by two jobs in runTheMatrix that run on already-produced AOD (136.7611* and 136.8311*) that fail because the needed collections are not found.

this should be fixed.
These re-mini workflows are using special era/process modifiers. Please use them to appropriately disable or perhaps instead produce "dummy" collections.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/1937/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
136.8311 step2
<pre>runTheMatrix-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017/step2_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017.log

</pre>136.7611 step2
<pre>runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/step2_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

slava77 commented:
>  Commits 63

@zucchett 
please also check if you can squash some intermediate commits to compactify the history a bit.

zucchett committed:
Added modifiers to exclude pixelClusterTagInfos from 80X and 94X miniAOD production

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11386

- This PR adds an extra 220KB to repository

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

cmsbuild commented:
+1
Tested at: 42c3ef49095fd92aeb02ba309be77fdfdf4650bd
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/1953/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/1953/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939165
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 142 log files, 14 edm output root files, 34 DQM output files

slava77 commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,87 @@
+    char L1_R004 = 0;
+    char L2_R004 = 0;
+    char L3_R004 = 0;
+    char L4_R004 = 0;
```
if this is to be used for integers, better use ```int8_t```

slava77 commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,87 @@
+    char L1_R008 = 0;
+    char L2_R008 = 0;
+    char L3_R008 = 0;
+    char L4_R008 = 0;
```
at least the layer items better be in an array

zucchett replied:
Ok, i implemented both changes in the latest commit.

zucchett committed:
PixelClusterTagInfos switched from char to int8_t and rearranged pixel layers in arrays

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11434

- This PR adds an extra 156KB to repository

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e51c6038bc584df17e971abf47b8426f09465d46
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2011/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2011/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 142 log files, 14 edm output root files, 34 DQM output files

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,162 @@
+      TVector3 c3(cluIt->x - firstPV->x(), cluIt->y - firstPV->y(), cluIt->z - firstPV->z());
+      TVector3 j3(jetRef->px(), jetRef->py(), jetRef->pz());
+      float dR = j3.DeltaR(c3);
+      float sC = 12. * 2. / (jetRef->pt());
```
Magic numbers?

perrotta commented on RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
```
Why this config file is needed?

perrotta commented on RecoBTag/PixelCluster/python/pixelCluster_cff.py:
```diff
@@ -0,0 +1,8 @@
+
+from RecoBTag.PixelCluster.pixelClusterTagInfos_cfi import *
+
+pixelClusterTask = cms.Task(
```
Where is this task used?
If not used, this config file is also useless

perrotta commented on RecoBTag/Configuration/python/RecoBTag_cff.py:
```diff
@@ -11,6 +11,7 @@
 from RecoBTag.TensorFlow.pfDeepDoubleX_cff import *
 from RecoBTag.MXNet.pfDeepBoostedJet_cff import *
 from RecoVertex.AdaptiveVertexFinder.inclusiveVertexing_cff import *
+from RecoBTag.PixelCluster.pixelCluster_cff import *
```
If you import here just directly "RecoBTag.PixelCluster.pixelClusterTagInfos_cfi", there is no need of this include of "RecoBTag.PixelCluster.pixelCluster_cff"

zucchett commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,60 @@
+  };
+
+  struct PixelClusterData {
+    int8_t R004[4] = {};
```
I like the option to pass the size of the array in the constructor, but in this case we can either use the new operator or switch to std::vectors. Which one would you prefer?

perrotta replied:
@zucchett , I honestly don't have preferences, provided memory is dealt with correctly

slava77 replied:
```std::array```  provides appropriate support for fixed size arrays

slava77 commented on RecoBTag/Configuration/python/RecoBTag_EventContent_cff.py:
```diff
@@ -32,6 +32,8 @@
         # DeepCMVA
         'keep *_pfDeepCMVAJetTags_*_*',
+        # Pixel Cluster
+        'keep *_pixelClusterTagInfos*_*_*',
```
we do not allow wildcards in the module names in the event content

slava77 commented on PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:
```diff
@@ -84,6 +84,8 @@
         # L1 prefiring weights
         'keep *_prefiringweight_*_*',
+        # Pixel Cluster
+        'keep *_pixelClusterTagInfos*_*_*',
```
here again please replace the module name wildcards

perrotta commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,60 @@
+    int8_t R010[4] = {};
+    int8_t R016[4] = {};
+    int8_t RVAR[4] = {};
+    unsigned int RVWT[4] = {};
```
CMS code rules require lowercase initials for class member variables: "r004", ..., "rVWT" (or "rvwt", or whatever you think it is most appropriate)

zucchett commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,162 @@
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (edmNew::DetSetVector<SiPixelCluster>::const_iterator detUnit = collectionClusters->begin();
```
It's certainly possible to use range loops, but since there are two nested loops we have to either:
1) make a copy of the DetSet<SiPixelCluster> object, slowing the code, or
2) have a not very readable code consisting of a chain of:
collectionClusters->data(d)->data(u) or even more complex syntax like
(*(*collectionClusters)[d])[u]

This is why we opted (and we still have the preference) for iterator loops in first place.

perrotta replied:
```C++
  for (auto & detUnit :  collectionClusters) {
```
wouldn't work?

zucchett replied:
Yes, but to make it work i have to first convert the edm::Handle<edmNew::DetSetVector<SiPixelCluster> >
to a
const edmNew::DetSetVector<SiPixelCluster>&

zucchett commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,162 @@
+    if (layer == 0 || layer > m_nLayers)
+      continue;
+
+    for (edmNew::DetSet<SiPixelCluster>::const_iterator clustIt = detUnit->begin(); clustIt != detUnit->end();
```
See also the comment above.

perrotta replied:
```C++
    for (auto& clustIt : detUnit) {
```
wouldn't work as well?

zucchett commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,162 @@
+    reco::PixelClusterData data = {};  // Initialize new Data to 0
+    reco::PixelClusterTagInfo tagInfo = {};
+
+    for (auto cluIt = clusters.begin(); cluIt != clusters.end(); ++cluIt) {
```
This one can certainly become a range based loop.

zucchett committed:
Several code improvements following PR's comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11488

- This PR adds an extra 160KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

zucchett commented:
The latest commit includes all the comments received so far.

perrotta commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,69 @@
+  private:
+    PixelClusterData pixelClusters;
+
+    edm::RefToBase<Jet> jetRef;
```
In CMSSW standards a trailing underscore indicates class members. Therefore, I'd rater rename
- ```jetRef``` -> ```jetRef_```
- ```pixelClusters``` -> ```pixelClusters_```

and above L48 and L54
```C++
    void setJetRef(const edm::RefToBase<Jet>& ref) { jetRef_ = ref; }
```
```C++
    void setData(const PixelClusterData& data) { pixelClusters_ = data; }
```

perrotta commented on RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
```diff
@@ -0,0 +1,89 @@
+  double m_minJetPt;
+  double m_maxJetEta;
+  double m_hadronMass;
+  double hadronMass;
```
By the way, there is no reason to have it "double", as it is only meant to build a "float"

perrotta commented on RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
```diff
@@ -0,0 +1,89 @@
+#include "TVector3.h"
+#include "TLorentzVector.h"
+
+class PixelClusterTagInfoProducer : public edm::stream::EDProducer<> {
```
Once you make all class member constant, I think this producer can even become a global producer

zucchett replied:
Thanks Andrea, i made the producer global, and merged the .h in the .cc. The other modifications have been applied, too. I just pushed the changes.

zucchett committed:
Moved .h in .cc, made producer global

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11506

- This PR adds an extra 164KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/BTauReco/src/classes.h  modified in PR(s): #27795
  - File  DataFormats/BTauReco/src/classes_def.xml  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/tools/jetTools.py  modified in PR(s): #27795
  - File  RecoBTag/Configuration/python/RecoBTag_cff.py  modified in PR(s): #27795

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -22,7 +22,89 @@
 
-#include "RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h"
+#ifndef RecoBTag_PixelCluster_PixelClusterTagInfoProducer_h
+#define RecoBTag_PixelCluster_PixelClusterTagInfoProducer_h
```
L25 and L26 not needed, since this file is not to be included anywhere

zucchett replied:
Ok, done!

perrotta reviewed (commented):
Thank you Alberto!
Just remove these two useless lines, and we can restart the tests

zucchett committed:
Removed ifndef define endif from PixelClusterTagInfoProducer.cc

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11508

- This PR adds an extra 172KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/BTauReco/src/classes.h  modified in PR(s): #27795
  - File  DataFormats/BTauReco/src/classes_def.xml  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #27795
  - File  PhysicsTools/PatAlgos/python/tools/jetTools.py  modified in PR(s): #27795
  - File  RecoBTag/Configuration/python/RecoBTag_cff.py  modified in PR(s): #27795

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 0cb01ddd87b6bb86b27726f13a8cdce268ca24cc
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2080/summary.html

cmsbuild commented:
Comparison job queued.

slava77 commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,69 @@
+  };
+
+  struct PixelClusterData {
+    std::vector<int8_t> r004;
```
it looks like this is used as an array of size 4.
Why was std::array not used?

slava77 replied:
ah, nm. there is also an explicit support for phase-0, aiming to save a few chars.

slava77 commented:
I was looking for some CPU/memory/storage measurements.
Something on PU\~50.
Did I miss it in this thread, or was it not done yet?

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -424,6 +424,32 @@ def miniAOD_customizeCommon(process):
+    from Configuration.Eras.Modifier_run2_miniAOD_94XFall17_cff import run2_miniAOD_94XFall17
+    run2_miniAOD_94XFall17.toReplaceWith(pfBTaggingTask, pfBTaggingTaskNoPixelCluster)
+    run2_miniAOD_94XFall17.toModify(process.patJets, addTagInfos = cms.bool(False) )
+    run2_miniAOD_94XFall17.toModify(process.slimmedJetsNoDeepFlavour, dropTagInfos = cms.string('1') )
```
here and other parts of this file/pr
please drop type specifications for all parameters which already exist.
This is a safer way to protect from parameter name mistakes and will also help in possible parameter migrations.

zucchett replied:
Type specifications have been dropped from this file (the other cfg being the pixelClusterTagInfos_cfi), except for InputTags.

zucchett replied:
Type specifications have been dropped, except for TagInfos and VTagInfos.

slava77 commented:
we usually do not create a new package for just one plugin/algorithm.
Can we use an existing package.
@fabiocos I see that you created a PR for the cms-bot to accept this package.
Was there a policy change?

zucchett commented:
Creating a new package was motivated by the fact that the new tagger being developed is very different from all other approaches to b-tagging, and therefore cannot be associated to any existing package. It's also likely that in the near future the tagger will be expanded and new tools and plugins will be added, thus requiring the creation of the new package anyway.

slava77 commented on DataFormats/BTauReco/src/classes_def.xml:
```diff
@@ -465,4 +465,10 @@
+  <class name="reco::PixelClusterData"/>
+  <class name="reco::PixelClusterProperties"/>
+  <class name="reco::PixelClusterTagInfo"/>
+  <class name="reco::PixelClusterTagInfoCollection"/>
```
all non-templated classes should have a class version for a more appropriate schema evolution tracking.

zucchett replied:
The class version (number 3) has been added to PixelClusterTagInfo and PixelClusterData.

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+
+// ROOT
+#include "TVector3.h"
+#include "TLorentzVector.h"
```
please switch to non-TObject vector variants.
See e.g.  in DataFormats/Math/interface

zucchett replied:
Ok, we replaced both TVector3 with GlobalPoints and reco::deltaR function.

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+      m_minADC(iConfig.getParameter<int>("minAdcCount")),
+      m_minJetPt((float)iConfig.getParameter<double>("minJetPtCut")),
+      m_maxJetEta((float)iConfig.getParameter<double>("maxJetEtaCut")),
+      m_hadronMass((float)iConfig.getParameter<double>("hadronMass")),
```
were these casts really needed?

zucchett replied:
They were introduced following a comment from @perrotta asking to make these values floats instead of double.

slava77 replied:
but the target values are already floats. The additional explicit cast does nothing.

zucchett replied:
Ok, i see. I removed the casts in the commit i just pushed.

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (auto& detUnit : collectionClusters) {
```
```auto const&```

zucchett replied:
Ok, we adopted the suggestion

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+    if (layer == 0 || layer > m_nLayers)
+      continue;
+
+    for (auto& clUnit : detUnit) {
```
```auto const&```

zucchett replied:
Same as before

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+    reco::PixelClusterData data(m_nLayers);
+    reco::PixelClusterTagInfo tagInfo;
+
+    for (auto& cluster : clusters) {
```
```auto const&```

zucchett replied:
Same as before

slava77 commented on RecoBTag/PixelCluster/python/pixelClusterTagInfos_cfi.py:
```diff
@@ -0,0 +1,14 @@
+import FWCore.ParameterSet.Config as cms
+
+pixelClusterTagInfos = cms.EDProducer("PixelClusterTagInfoProducer",
```
this file is redundant, mostly.
```isPhase1``` should be set appropriately via an era

zucchett replied:
We added the era modifier from Configuration/Eras/python/Modifier_pixel_2016_cff

slava77 commented on PhysicsTools/PatAlgos/python/recoLayer0/bTagging_cff.py:
```diff
@@ -202,6 +204,8 @@
   , 'pfMassIndependentDeepDoubleCvBJetTags:probHbb'                     : [["pfDeepDoubleXTagInfos"], ['pfBoostedDoubleSVAK8TagInfos', "pfImpactParameterAK8TagInfos", 'pfInclusiveSecondaryVertexFinderAK8TagInfos']]
   , 'pfMassIndependentDeepDoubleCvBJetTags:probHcc'                     : [["pfDeepDoubleXTagInfos"], ['pfBoostedDoubleSVAK8TagInfos', "pfImpactParameterAK8TagInfos", 'pfInclusiveSecondaryVertexFinderAK8TagInfos']]
+  # Pixel Cluster tag infos
+  , 'pixelClusterBJetTags'                               : [['pixelClusterTagInfos']]
```
what does this refer to ? I do not see a ```pixelClusterBJetTags``` defined

zucchett replied:
This line can be safely deleted.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -424,6 +424,32 @@ def miniAOD_customizeCommon(process):
+
+    # Embed pixelClusterTagInfos in slimmedJets
+    process.patJets.addTagInfos = cms.bool(True)
+    process.patJets.tagInfoSources.append( cms.InputTag("pixelClusterTagInfos") )
```
shouldn't this be done with the standard btag tools? PhysicsTools/PatAlgos/python/tools/jetTools.py

zucchett replied:
A line in jetTools.py to include pixelClusterTagInfos has been added since the beginning (L595 in jetTools.py)
In principle, it's not necessary to use the jetTool or perform additional updateJetCollection steps: everything we need is already there, we just need to set the existing modules to keep the pixelClusterTagInfos and not discard them.

ferencek replied:
This is somewhat in collision with changes made in ` PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py`. How do you plan to access these TagInfos later on? If you plan to access them directly, then adding `'keep *_pixelClusterTagInfos_*_*'` is fine and you don't need to set `dropTagInfos = '0'` and `addTagInfos = True`. In fact, all of this addition of TagInfos is for nothing because they are later dropped by the output module because there is no explicit keep statement for them. However, if they were kept, there would be a big jump in the event size because you are not only adding pixelClusterTagInfo (note the `tagInfoSources.append( cms.InputTag("pixelClusterTagInfos") )` call) but all other TagInfos.

If you really need to add TagInfo to slimmed jets, keep in mind that they are really stored in a separate collection (the name starts with `recoBaseTagInfosOwned_...`) so in that case you would have to ensure that that particular collection of TagInfos is kept, that you only keep pixel cluster TagInfos (so no appending) and you would also need to implement an appropriate TagInfo accessor method in `pat::Jet` similar to those in https://github.com/cms-sw/cmssw/blob/CMSSW_10_6_2/DataFormats/PatCandidates/interface/Jet.h#L198-L210.

zucchett replied:
Thanks Dinko for the suggestion. Indeed, the plan is to use the second method (be able to  get the Tag Infos from slimmedJets). In the latest commit, we drop `*_pixelClusterTagInfos_*_*`, but we now keep `recoBaseTagInfosOwned_slimmedJets_*_*`, replacing the `tagInfoSources.append( cms.InputTag("pixelClusterTagInfos") )` with `tagInfoSources = cms.VInputTag( cms.InputTag("pixelClusterTagInfos") )` in `miniAOD_tools.py`. We also implemented the new accessor method in Jet.h and Jet.cc.

slava77 commented on RecoBTag/Configuration/python/RecoBTag_cff.py:
```diff
@@ -35,7 +36,10 @@
+    combinedMVAV2BJetTags,
+    
+    # pixel cluster
+    pixelClusterTagInfos,
```
BTW, what is the meaning of "legacy" here and why is newly added ```pixelClusterTagInfos``` expected to belong to it?

zucchett replied:
The legacy b-tagging sequence has been there a long time before this PR, and my understanding (BTV experts may want to comment) is that the difference with the standard sequence is the use of PF-based taggers vs non-PF. PixelClusters do not rely on PF but on low level information, so it should be the same for both PF and non-PF.

slava77 commented on RecoBTag/Configuration/python/RecoBTag_cff.py:
```diff
@@ -79,3 +85,17 @@
+
+## FastSim modifier (do not run pixelClusterTagInfos because the siPixelCluster input collection is not produced in FastSim)
+from Configuration.Eras.Modifier_fastSim_cff import fastSim
+fastSim.toReplaceWith(btagging, btaggingNoPixelCluster)
```
wouldn't it be easier to short-circuit the ```pixelClusterTagInfos``` module configuration than manipulate the sequences?
Also, if these \*NoPixelCluster variants are not meant to be exported, better prefix with ```_```

zucchett replied:
Not sure what short-circuit would mean here. The *NoPixelCluster are actually exported, for instance in miniAOD_tools.py for Era modifiers. Simply adding the underscore instead of  NoPixelCluster would make the name less clear (in my opinion).

perrotta replied:
@zucchett , I think that the suggestion from @slava77 was to apply the modifiers to the pixelClusterTagInfos module, and not to the main reco BTag sequence (or miniAOD).
You can use modifiers to replace pixelClusterTagInfos with some dummy/empty module inside RecoBTag/PixelCluster/python/pixelClusterTagInfos_cfi.py, and you don't have to apply modifiers elsewhere

zucchett replied:
Ok, i can short-circuit the PixelCluster producer by setting a very high threshold on jet pT, so an empty collection is produced and does not fail because it never looks for the siPixelCluster collection. I moved the modifiers to pixelClusterTagInfos_cfi.py, and removed them from the b-tagging sequence and partially from miniAOD_tools.py. However in the latter i still have to avoid that the code looks for the pixelClusterTagInfo collection, so some modifiers on the modules in the miniAOD sequnce are still needed.

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+
+    if (subid == 1) {  // pixel barrel
+      PixelBarrelName pbn(detId, tTopo, m_isPhase1);
+      layer = pbn.layerName();
```
Isn't the TrackerTopology enough?
```pxbLayer``` and ```pxfDisk``` accessors are available .

zucchett replied:
The pixel layer is accessed directly from TrackerTopology::layer(), which uses pxbLayer and pxfDisk.

slava77 commented:
>  and BTV conveners @mverzett , @kskovpen, @XavierAtCERN

also @ferencek 

was this code also reviewed by the BTV group?
At a glance, not everything looks like it follows other btag setup practices.

zucchett commented on DataFormats/BTauReco/interface/PixelClusterTagInfo.h:
```diff
@@ -0,0 +1,69 @@
+    edm::RefToBase<Jet> jet() const override { return jetRef_; }
+
+    // method to set the PixelClusterData
+    void setData(const PixelClusterData& data) { pixelClusters_ = data; }
```
We actually use them in the TagInfo Producer, and we would like to keep the possibility to update the jet ref and the data object separately.

zucchett commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,262 @@
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
```
The /stream line has been removed.

zucchett commented:
Tests have been performed on TTbar RelVal samples with simulated PU (/store/relval/CMSSW_10_5_0_pre2/RelValTTbar_13/GEN-SIM-RECO/PU25ns_105X_upgrade2018_realistic_v2-v1) by running the full b-tag sequence without and with the pixelClusterTagInfoProducer:

Peak virtual memory size went from 1966.84 Mbytes to 1967.13 Mbytes (<0.1% increase).

CPU total job increase is from 266.05 to 268.205 s, but even larger variations are observed by repeating the measurement.

Average processing time of the PixelClusterTagInfoProducer with TimeModule over 100 events: 0.0030 s.

zucchett committed:
Several updates with latest comments from PR

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11535

- This PR adds an extra 176KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

zucchett committed:
Removed unnecessary explicit cast on input parameters

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11538

- This PR adds an extra 176KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
> Average processing time of the PixelClusterTagInfoProducer with TimeModule over 100 events: 0.0030 s.

perhaps an igprof would be a better measure.

zucchett commented:
IgProf reports, although very lengthy, are here:
https://zucchett.web.cern.ch/zucchett/Pixel/IgProf/igreport_perf_pixel.res (full b-tag sequence)
https://zucchett.web.cern.ch/zucchett/Pixel/IgProf/igreport_perf.res (only PixelClusterTagInfo Producer)
the TimeModule results, filtered only for the PixelClusterTagInfo to be interpreted more easily, are here:
https://zucchett.web.cern.ch/zucchett/Pixel/IgProf/times.txt
reporting the time for each event. The results can also be analyzed in this spreadsheet:
https://docs.google.com/spreadsheets/d/1ZodegFYdsymzVAabkHpDBiYHdx1YxTk7gX7RHtCIFEE/edit?usp=sharing
The ~50% fraction of event with a very small time correspond to the events with no central jets with pT>100 GeV.
![pixelClusterTagInfoProducer](https://user-images.githubusercontent.com/6450143/63538938-3d9c0480-c519-11e9-8881-d845a7665376.png)

slava77 commented:
On 8/22/19 11:13 AM, Alberto Zucchetta wrote:
> IgProf reports, although very lengthy, are here:

please make a browsable version.
https://twiki.cern.ch/twiki/bin/viewauth/CMS/RecoIntegration#Run_profiler_igprof
Thank you.

zucchett commented:
I added the browsable versions. For CPU use:
http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_perf
and memory use:
http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_total

zucchett committed:
Restored charge collection as unsigned int (lost in previous commits)

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11567

- This PR adds an extra 172KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

slava77 commented:
>  For CPU use:
> http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_perf

thank you for the link.
Please point me to the config.py or cmsDriver command used to run this.
I see that only one module shows up in the profile.

slava77 commented:
from http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_perf/477
it looks like we need at least an order of magnitude more events processed to have a good measurement.

zucchett commented:
I copied the config here:
https://zucchett.web.cern.ch/zucchett/Pixel/IgProf/testRun.py
i'm re-running the memory tests with 1000 events.

slava77 commented:
> I copied the config here:
> https://zucchett.web.cern.ch/zucchett/Pixel/IgProf/testRun.py
> i'm re-running the memory tests with 1000 events.

which release is this?
I see that you have ```auto:run2_mc_FULL``` which is nebulously pointing to ```phase1_2021_realistic``` since 1100pre3 (run3 global tag), compared to the ```105X_upgrade2018_realistic_v2``` used to make the test file.

zucchett commented:
I am currently using a CMSSW_11_0_0_pre5. I'll switch to the `105X_upgrade2018_realistic_v2` to rerun the tests.

slava77 commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+
+    // Get the geom-detector
+    const PixelGeomDetUnit* geomDet = dynamic_cast<const PixelGeomDetUnit*>(theTracker.idToDet(detId));
+    const PixelTopology* topol = &(geomDet->specificTopology());
```
it seems like these casts are not needed. IIUC, all methods are virtual and are available in the base classes.
So, the following should supposedly work
```C++
const auto* geomDet = theTracker.idToDet(detId);
const auto* topol = &geomDet->topology();
```

zucchett replied:
Ok, no error is raise during compilation or runtime. We implemented this in the latest commit.

zucchett committed:
Modified miniAOD sequence to access TagInfos from recoBaseTagInfosOwned_* collection

zucchett committed:
Defined dedicated accessor methods for PixelClusterTagInfo

cmsbuild commented:
The code-checks are being triggered in jenkins.

zucchett commented:
In the latest commits, we implemented Dinko's suggestion (BTV review) to handle and access the tagInfos in miniAODs, and updated the igprof tests with more events and the latest comments from @slava77 . The links are the same:
http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_perf
http://zucchett.web.cern.ch/zucchett/cgi-bin/igprof-navigator/igreport_total

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11615

- This PR adds an extra 192KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented on PhysicsTools/PatAlgos/python/slimming/MicroEventContent_cff.py:
```diff
@@ -11,8 +11,8 @@
-        # drop content created by MINIAOD DeepFlavour production
-        'drop recoBaseTagInfosOwned_slimmedJets_*_*',
+        # keep slimmedJets TagInfos, currently only PixelClusterTagInfo
+        'keep recoBaseTagInfosOwned_slimmedJets_*_*',
```
This line is redundant, as it is fully contained in L13, but maybe it can be left as a reminder for the presence of those TagInfos.
@ferencek please anyhow have a look and confirm whether this satisfies your comment.

ferencek replied:
Thinks look ok to me now. Indeed, L15 is redundant but I agree it probably doesn't hurt to leave it along with L14 as a reminder.

zucchett committed:
Moved modifiers from RecoBTag_cff and partially from miniAOD_tools.py to pixelClusterTagInfos_cfi

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11624

- This PR adds an extra 208KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7329c23c5ec7d00464d3011a041cc52d44b87aa0
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2181/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2181/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

ferencek commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -424,6 +424,25 @@ def miniAOD_customizeCommon(process):
+    process.patJets.tagInfoSources = cms.VInputTag( cms.InputTag("pixelClusterTagInfos") )
+    process.slimmedJetsNoDeepFlavour.dropTagInfos = '0'
+    process.updatedPatJetsSlimmedDeepFlavour.addTagInfos = True
+    process.updatedPatJetsSlimmedDeepFlavour.tagInfoSources = cms.VInputTag( cms.InputTag("pixelClusterTagInfos") )
```
Looking at it more carefully, any changes to `updatedPatJetsSlimmedDeepFlavour` are not needed because TagInfos get added to `updatedPatJetsTransientCorrectedSlimmedDeepFlavour`.

zucchett replied:
Thanks @ferencek , indeed changes to `updatedPatJetsSlimmedDeepFlavour` do not have any effect. They have been removed now also from these lines and the era modifiers, few lines below.

zucchett committed:
Removed unnecessary changes on updatedPatJetsSlimmedDeepFlavour for pixelClusterTagInfo embedding

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11665

- This PR adds an extra 216KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1130a414f8ba5733f5bcd0a341151daea2294bba
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2228/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2228/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 2
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939165
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

perrotta commented:
I evaluated the extra size and extra cpu time produced by this PR on a (small) sample of 100 events produced with wf 11024 (2018 TTbar with PU).

- Event size in reco output, 70 B/evt on average:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->        69.7         70     NEWO   0.00     recoPixelClusterTagInfos_pixelClusterTagInfos__RECO.
-------------------------------------------------------------
  3525365 ->     3525425         60             0.0     ALL BRANCHES
```

- Event size miniAOD output, also about 70 B/evt on average:
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
      0.0 ->        67.5         67     NEWO   0.09     recoBaseTagInfosOwned_slimmedJets_tagInfos_RECO.
-------------------------------------------------------------
    78137 ->       78211         74             0.1     ALL BRANCHES
```

- CPU timing:
The new pixelClusterTagInfos module adds about 3 ms/evt to the processing time, which represents 0.02% of the total reco + miniAOD processing time (no validation, dqm steps):
```
The same excluding the first 1 events
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
       added      +0.02%         0.00 ms/ev ->         2.95 ms/ev pixelClusterTagInfos
  ---------- ------------     --------                  ----       ------------
Job total:  15.7006 s/ev ==> 15.2512 s/ev
```

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+
+// Pixel topology
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/SiPixelDetId/interface/PixelBarrelName.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+// Pixel topology
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/SiPixelDetId/interface/PixelBarrelName.h"
+#include "DataFormats/SiPixelDetId/interface/PixelEndcapName.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+#include "DataFormats/SiPixelDetId/interface/PixelSubdetector.h"
+#include "DataFormats/SiPixelDetId/interface/PixelBarrelName.h"
+#include "DataFormats/SiPixelDetId/interface/PixelEndcapName.h"
+#include "DataFormats/SiPixelDetId/interface/PXBDetId.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+#include "DataFormats/SiPixelDetId/interface/PixelBarrelName.h"
+#include "DataFormats/SiPixelDetId/interface/PixelEndcapName.h"
+#include "DataFormats/SiPixelDetId/interface/PXBDetId.h"
+#include "DataFormats/SiPixelDetId/interface/PXFDetId.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+// Geometry
+#include "DataFormats/Math/interface/deltaR.h"
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+#include "DataFormats/GeometryVector/interface/GlobalPoint.h"
+#include "Geometry/Records/interface/IdealGeometryRecord.h"
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetUnit.h"
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetType.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+#include "Geometry/TrackerGeometryBuilder/interface/PixelGeomDetType.h"
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/CommonDetUnit/interface/GeomDetType.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+#include "Geometry/TrackerGeometryBuilder/interface/TrackerGeometry.h"
+#include "Geometry/Records/interface/TrackerDigiGeometryRecord.h"
+#include "Geometry/CommonDetUnit/interface/GeomDetType.h"
+#include "Geometry/CommonTopologies/interface/PixelTopology.h"
```
This include doesn't seem to be needed

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+      continue;
+    DetId detId = DetId(detUnit.detId());  // Get the Detid object
+    unsigned int detType = detId.det();    // det type, pixel = 1
+    if (detType != 1)
```
"1" here appears like a magic number. Please replace L166 and L167 with the single
```C++
    if (detId.det() != DetId::Tracker)
```

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+      // get global position of the cluster
+      LocalPoint lp = topol->localPosition(MeasurementPoint(clUnit.x(), clUnit.y()));
+      GlobalPoint clustgp = geomDet->surface().toGlobal(lp);
+      if (m_minADC > 0 and clUnit.charge() < m_minADC)
```
Please move at the beginning of the loop: no reason to compute ```lp``` and ```clustgp``` if below the adc threshold

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+    edm::RefToBase<reco::Jet> jetRef = collectionJets->refAt(j);  // Get jet RefToBase
+
+    reco::PixelClusterData data(m_nLayers);
+    reco::PixelClusterTagInfo tagInfo;
```
Move this declaration where it is used, i.e. right before line 227

zucchett replied:
Ok, and i can pass the jet ref and the data object to the constructor.

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+
+    for (auto const& cluster : clusters) {
+      GlobalPoint c3(cluster.x, cluster.y, cluster.z);
+      GlobalPoint v3(firstPV->x(), firstPV->y(), firstPV->z());
```
```v3``` can be computed only once outside these two for loops, either before L192 or right after the definition of ```firstPV``` at L136

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+      GlobalPoint c3(cluster.x, cluster.y, cluster.z);
+      GlobalPoint v3(firstPV->x(), firstPV->y(), firstPV->z());
+      float dR = reco::deltaR(c3 - v3, jetRef->momentum());
+      float sC = m_hadronMass * 2. / (jetRef->pt());  // 2 mX / pT
```
This doesn't need to be recomputed for every cluster: please move outside this loop

perrotta commented on RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc:
```diff
@@ -0,0 +1,253 @@
+      // Match pixel clusters to jets and fill Data struct
+      if (cluster.layer >= 1 && cluster.layer <= m_nLayers) {
+        int idx(cluster.layer - 1);
+        if (dR < 0.04)
```
Some unneeded computations (although admittedly risible in terms of cpu time) could be saved if these conditions are rewritten as:
```C++
        if (dR < 0.16) {
          data.r016[idx]++;
	  if (dR < 0.10) {
	    data.r010[idx]++;
	    if (dR < 0.08) {
	      data.r008[idx]++;
	      if (dR < 0.06) {
		data.r006[idx]++;
		if (dR < 0.04)
		  data.r004[idx]++;
	      }
	    }
	  }
	}
        if (dR < sC) {
          data.rvar[idx]++;
          data.rvwt[idx] += cluster.charge;
	}
```

perrotta reviewed (commented):
I tested the following suggestions and the cpu time seems to shrink from 2.95 ms/evt to 2.75 ms/evt , although the difference could be also partially due to the resolution of the method.  In any case applying these updates doesn't hurt

zucchett committed:
PixelClusterTagInfo code improvements and cleanup

zucchett commented:
In the latest commit, i implemented all latest @perrotta suggestions.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27738/11688

- This PR adds an extra 224KB to repository

- Found files with invalid states:
  - RecoBTag/PixelCluster/python/pixelCluster_EventSetup_cff.py:
    - Added: beb22fc42590471ad68042aac5cfbd507626616b, a1d5cf4c068ed2fd34b30fe0c3e0e619638c2b03
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/python/pixelCluster_cff.py:
    - Added: b9bc8c88b0c3bf4ea664e3a976746f4d12515af1, bd745e3c1d07221e5f90713f5bfe471c9abe2ca0
    - Modified: d0e05d59bcd06ea7e92649c44439ea34b2e20e02, 54ab000a59023c2cb88e89ef9c2878f61ae6b570
    - Deleted: 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
  - RecoBTag/PixelCluster/interface/PixelClusterTagInfoProducer.h:
    - Added: 8bfd33e8907809f1206ddd24377080b1e4d7c95f, 183ef587c8b6b3705c14b1b299ffc512f534c4df
    - Modified: 19d913d198a5200c981713c3deb4ab99532c99dc, a793da03fa97008ca79f7528ecb4e9f9e24e0dbf, d48f6695ec997bc9f559ca3a58e89e2f6131bccd, c7e4961d3259d2ca50f9f502ccb57d0e14e79013, b483ae6ff31f6a314e65cc14dcb31e3efd2c73fc, 5b460d80c0aff1e24e2a19f68c2fa2e48058cf9b
    - Deleted: 409e8a34ed623a3843826b2fb1e9c4da7a238c06

cmsbuild commented:
Pull request #27738 was updated. @perrotta, @cmsbuild, @santocch, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 83310006eeccb09e2cf84320057013ce29ce83a9
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2247/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-343cde/2247/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2939508
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2939166
 - DQMHistoTests: Total skipped: 341
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 145 log files, 15 edm output root files, 34 DQM output files

perrotta commented:
Thank you @zucchett , PR code is fine now for reco,

Before finally signing it, could you please verify that the output does correspond to what you are expecting? You can check any root file already produced by the automatic tests, for example.

As far as I can see, we're getting:
- recoPixelClusterTagInfos_pixelClusterTagInfos added to reco/aod content
- recoBaseTagInfosOwned_slimmedJets (new one) and recoBaseTagInfosOwned_slimmedJetsPuppi are saved into the miniAOD event content, according to what specified in MicroEventContent_cff.py

In the reco/aod output I can see the vectors in pixelClusters_ with some reasonable content.
Since it is not obvious to me how to access the tagging variables in miniAOD, I'd let you verify that they correctly contain all what is needed

zucchett commented:
Thanks @perrotta , i verified that i can access the collection in RECO/AOD and the jet tag Infos from slimmedJets in miniAOD, and get consistent results.

In miniAOD, `recoBaseTagInfosOwned_slimmedJets` is correctly added to the event content, but on the other hand, the `recoBaseTagInfosOwned_slimmedJetsPuppi` in miniAOD was not explicitly introduced by us (PixelCluster TagInfos cannot be neither matched nor recalculated for Puppi jets). I believe it was already included because of the pre-existing `'keep *_slimmedJetsPuppi_*_*'` command.

As long as this is ok for you, you can proceed with this PR.

perrotta commented:
+1
- recoPixelClusterTagInfos_pixelClusterTagInfos produced and added to reco/aod event content in accordance with what planned in the PR descritpion and adapted to the outcome of the review
- new recoBaseTagInfosOwned_slimmedJets propagates this info into the miniAOD event content
- Jenkins tests show that other reco products are not affected
- pixelClusterTagInfos are not included in FastSim, nor in the eras corresponding to past campaigns

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

peruzzim commented:
of course XPOG is ok with the (very small) addition to MINIAOD event content

kpedro88 commented:
+1
